rules/when-to-use-domain-and-integration-events/rule.md

```diff
@@ -0,0 +1,61 @@
+---
+type: rule
+title: Do you know when to use Domain and Integration Events?
+guid: e91f22d2-d385-49b6-a27b-1195b4bb13f9
+uri: when-to-use-domain-and-integration-events
+created: 2023-07-25T11:01:35.0000000Z
+authors:
+- title: William Liebenberg
+  url: https://ssw.com.au/people/william-liebenberg
+archivedreason:
+related: []
+redirects:
+---
+
+Domain Events and Integration Events are powerful patterns used in Domain-Driven Design (DDD) to enable loosely coupled communication between different parts of an application. They serve as a means of **notification** for important domain concepts that have occurred.
+
+<!--endintro-->
+
+## What are Domain Events
+
+In the Domain Layer of the [Clean Architecture](https://github.com/sswconsulting/SSW.CleanArchitecture), Domain Events are an integral part of the Domain Model.
+
+Domain Events are raised within the Domain Layer of your application when an entity or aggregate makes a significant decision or undergoes a state change. The main purpose of Domain Events is to enable loose coupling and keep domain logic isolated from the application's infrastructure.
+
+For example, when an `Order` is placed, a domain event can be raised (e.g. `OrderReceivedEvent`) to notify other parts of the domain that need to react to this event, such as updating inventory or sending confirmation emails.
+
+Another example could be when the `Order`'s `Status` changes from `OrderStatus.Received` to `OrderStatus.Processing`, we can publish an `OrderStatusChanged` event.
+
+It's essential to identify suitable events as not everything qualifies as a Domain Event. For instance, "client walked into the store" or "chicken crossed the road" may not be appropriate for a Domain Event.
+
+Domain events are immutable and can be considered as **historical facts** capturing something that occurred in the domain process. They are meant to be a representation of past events and cannot be altered or disputed.
+
+When describing domain events, we commonly use a past-tense naming convention, such as `OrderCreated`, `UserRegistered`, `InvoiceConsolidated`.
+
+Domain events should not depend on external dependencies or external systems, adhering to the principles of the Domain Layer.
+
+When publishing a domain event, the entire Entity or Aggregate can be included since the event's scope is confined to the current bounded context.
+
+Note: It is important to remember that the definition and behavior of a `Product` in one bounded context might differ from another bounded context, like an e-commerce application product versus the product of a chemical reaction in a laboratory.
```

Comment:
I don't think this note adds anything...could take it out. If we keep it, suggest putting it in an info box.

Reply:
Think of, we only return DTOs from the commands/queries... we don't return entities. Same idea that the domain events that contain whole entities don't go beyond the app/infra layer since it might not be well understood outside of the application.

But yes, i think an info box would look nicer :)